src/library/scala/collection/immutable/LazyList.scala

```diff
@@ -258,51 +258,89 @@ import scala.runtime.Statics
   *  @define evaluatesAllElements This method evaluates all elements of the collection.
   */
 @SerialVersionUID(3L)
-final class LazyList[+A] private(private[this] var lazyState: () => LazyList.State[A])
+final class LazyList[+A] private (lazyState: () => LazyList[A])
   extends AbstractSeq[A]
     with LinearSeq[A]
     with LinearSeqOps[A, LazyList, LazyList[A]]
     with IterableFactoryDefaults[A, LazyList]
     with Serializable {
   import LazyList._
 
-  @volatile private[this] var stateEvaluated: Boolean = false
-  @inline private def stateDefined: Boolean = stateEvaluated
-  private[this] var midEvaluation = false
-
-  private lazy val state: State[A] = {
-    // if it's already mid-evaluation, we're stuck in an infinite
-    // self-referential loop (also it's empty)
-    if (midEvaluation) {
-      throw new RuntimeException(
-        "LazyList evaluation depends on its own result (self-reference); see docs for more info"
-      )
-    }
-    midEvaluation = true
-    val res = try lazyState() finally midEvaluation = false
-    // if we set it to `true` before evaluating, we may infinite loop
-    // if something expects `state` to already be evaluated
-    stateEvaluated = true
-    lazyState = null // allow GC
-    res
+  private def this(head: A, tail: LazyList[A]) = {
+    this(LazyList.EmptyFun)
+    _head = head
+    _state = tail
   }
 
+  // used to synchronize lazy state evaluation
+  // when `_head ne Uninitialized`
+  //   - `null` if this is an empty lazy list
+  //   - `head: A` otherwise (can be `null`, `_state == null` is used to test emptyness)
+  @volatile private[this] var _head: Any = Uninitialized
+
+  // when `_head eq Uninitialized`
+  //   - `lazySate: () => State`
+  //   - MidEvaluation while evaluating lazyState
+  // when `_head ne Uninitialized`
+  //   - `null` if this is an empty lazy list
+  //   - `tail: LazyList[A]` otherwise
+  private[this] var _state: AnyRef = lazyState
+
+  private def rawHead: Any = _head
+  private def rawState: AnyRef = _state
+
+  @inline private def stateDefined: Boolean = _head.asInstanceOf[AnyRef] ne Uninitialized
+
+  private def state: LazyList[A] =
+    if (stateDefined) {
+      if (_state == null) LazyList.Empty
+      else this
+    } else {
+      synchronized {
```

Comment:
Should the entire `else` branch be moved a separate method? We do that for `lazy val`s, so that the `def state` accessor is small enough to be inlined by the JVM.